src/transform/src/predicate_pushdown.rs

```diff
@@ -607,6 +619,69 @@ impl PredicatePushdown {
         }
     }
 
+    /// If `s` is of the form
+    /// `(isnull(expr1) && isnull(expr2)) || (expr1 = expr2)`,
+    /// extract `expr1` and `expr2`.
+    fn extract_equal_or_both_null(s: &mut MirScalarExpr) -> Option<(MirScalarExpr, MirScalarExpr)> {
+        // Or, And, and Eq are all commutative functions. For each of these
+        // functions, order expr1 and expr2 so you only need to check
```

Comment:
This code demonstrates some serious self-awareness, I would not want to be in the same room with it. For example, in this branch:

```
materialize=> explain  SELECT  *  FROM t1 RIGHT JOIN t2 ON (TRUE) WHERE t1.f1 IS NULL AND t1.f1 + t1.f2 IS NOT NULL;
 Optimized Plan 
----------------
 %0 =          +
 | Constant    +
```
Which is indeed the case, as the ```WHERE``` clause is impossible. However, this only happens with joins. With single tables the query is not judged to be impossible right away:

```
materialize=> explain  SELECT  *  FROM t1  WHERE  t1.f1 IS NULL AND t1.f1 + t1.f2 IS NOT NULL;
           Optimized Plan           
------------------------------------
 %0 =                              +
 | Get materialize.public.t1 (u49) +
 | Filter !(isnull(#0)), isnull(#0)+
```